### PR TITLE
Moving IDeferredTracerProviderBuilder to API

### PR DESCRIPTION
**Pull request by CodeBlanch**
Related to #1889, #2058

## Changes

Moved IDeferredTracerProviderBuilder from SDK to API. The goal is to allow instrumentation libraries to be written without a dependency on SDK.

## Public API

API doesn't have a dependency on Microsoft.Extensions.DependencyInjection so IDeferredTracerProviderBuilder can no longer expose `IServiceCollection`.

```csharp
// Moved from OpenTelemetry to OpenTelemetry.Api
namespace OpenTelemetry.Trace
{
    public interface IDeferredTracerProviderBuilder
    {
        // removed
        // IServiceCollection Services { get; }

        // existing
        TracerProviderBuilder Configure(Action<IServiceProvider, TracerProviderBuilder> configure);
    }
}

// Added to OpenTelemetry.Extensions.Hosting
namespace OpenTelemetry.Trace
{
    public static class TracerProviderBuilderExtensions
    {
       public static IServiceCollection GetServices(this TracerProviderBuilder tracerProviderBuilder) {}
    }
}
```

## TODOs

* [X] `CHANGELOG.md` updated for non-trivial changes
* [ ] Changes in public API reviewed

### CODE REVIEW

**Review by cijothomas — APPROVED** [review]
LGTM!

**Comment by cijothomas** [comment]
Merging quickly as this was discussed in the SIG meetings, and this is required for 1.0.0-beta5 release, (and to unblock 1.1.0 as well).
If there are any additional feedback on this, please leave a comment! Thanks.

**Comment by danielcweber** [comment]
Would you be open to have an interface, let's say `IFancyDeferredTracerProviderBuilder` that would live in the SDK and again expose `IServiceCollection`? Code with a reference to the SDK could type check `TracerProviderBuilder` on the interface (like it's often done today) to get a hold of an `IServiceCollection`. Background: I'm instrumenting some stuff, and mostly it could be done be decorating existing DI-registrations....but no way to do it in one place (by a single extension method of `TracerProviderBuilder` since, well, no `IServiceCollection`). I'm not sure how all the instrumentations can easily live without decorators.

**Comment by CodeBlanch** [comment]
@danielcweber I don't think anything is off the table 😄 I'm not totally following you though. Could you post some pseudo-code (or whatever) to kind of demo how this would be used?

One potential solution you can do right now...

This extension allows you to register a callback action which will be fired when the `TraceProvider` build is invoked:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/8c1c1495d90bbf9d9cf6ba0a2a878c7aa1b1155a/src/OpenTelemetry.Extensions.Hosting/Trace/TracerProviderBuilderExtensions.cs#L90

Now of course you need the `TracerProviderBuilder` instance to do that, so it isn't immediately helpful for your case, but it can still be helpful! Let's say you define some helper service you can register like...

```csharp
public delegate void ConfigureTracerProviderAction(IServiceProvider serviceProvider, TracerProviderBuilder builder);
```

Now we can use that extension above to (at "build" time) find extra external configurations and run them:

```csharp
            services.AddOpenTelemetryTracing((builder) => builder
                .Configure((sp, builder) =>
                {
                    IEnumerable<ConfigureTracerProviderAction> registeredConfigurations = sp.GetServices<ConfigureTracerProviderAction>();
                    foreach (ConfigureTracerProviderAction registeredConfiguration in registeredConfigurations)
                    {
                        registeredConfiguration(sp, builder);
                    }
                }));
```

That allows you to then register anything you want to run with the `IServiceCollection` independently of `TracerProviderBuilder` (so long as you can access `IServiceCollection`):

```csharp
            services.AddTransient<MyInstrumentation>();
            services.AddSingleton<ConfigureTracerProviderAction>((sp, builder) => builder
                .AddInstrumentation(() => sp.GetRequiredService<MyInstrumentation>()));
```

**Comment by danielcweber** [comment]
>>That allows you to then register anything you want to run with the IServiceCollection independently of TracerProviderBuilder (so long as you can access IServiceCollection):

That's exactly the point. I don't want to have to configure something on `TracerProviderBuilder` *and* at the same time register services on an `IServiceCollection`. I want my instrumentation to be configured by a single call to an extension to `TracerProviderBuilder`. But since there's a need to register decorators, I need to get an `IServiceCollection` from the `TracerProviderBuilder`. Its usually in there, since `TracerProviderBuilderHosting` wraps it...but there's no interface that would expose it.

Basically, I want to repeat the pattern [seen here](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.InMemory/InMemoryExporterHelperExtensions.cs#L38), but with an interface that would let me get the `IServiceCollection`.

**Comment by CodeBlanch** [comment]
@danielcweber 

If you just want to make an extension method that configures + does service registrations, does this work?

https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Extensions.Hosting#building-extension-methods

You have to take a dependency on OpenTelemetry.Extensions.Hosting to get access to all the fun extensions.

`GetServices` gives you back the `IServiceCollection`:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/8c1c1495d90bbf9d9cf6ba0a2a878c7aa1b1155a/src/OpenTelemetry.Extensions.Hosting/Trace/TracerProviderBuilderExtensions.cs#L107-L115

**Comment by danielcweber** [comment]
Awesome. I had no clue this extension existed. I should have searched for it more thoroughly.

Thanks a lot. That should do!
